### PR TITLE
Entry to the challenge based on a optimizing a neural network on S/N metric

### PR DESCRIPTION
**Pull request by EiffL**
Here is my proposal for how to solve the challenge :-D in one word: autodiff \o/

Intrinsically the thing that we want to do is to find an optimal bin attribution rule by direct optimization of
the metric. Classically, this is intractable, because to do it efficiently you would have to get derivatives through the metric, which is coming out of CCL. So.... what to do???? 

Well, thanks to the confinement, I've finally found time to write an **end-to-end differentiable and GPU accelerated cosmology library**.

Introducing `jax-cosmo` : https://github.com/DifferentiableUniverseInitiative/jax_cosmo

I have re-implemented all of the machinery to get from cosmology parameters to 3x2pt angular Cls by Limber in the JAX framework. JAX is like TensorFlow or PyTorch, but following the NumPy API, so essentially, your code looks like plain old numpy code, except, it runs on GPU, and is automatically differentiable: JAX = NumPy + GPU + Autodiff 

Sooo, what this PR proposes, is a `jax-cosmo` re-implementation of the challenge metric, and now that the metric is differentiable, you can put in the middle a neural network to assign bins given photometry, and optimize!

So what happens is that we by-pass the step of photo-z estimation, and directly assign a bin number, just in order to maximize S/N.  

And it works !!!!

Here is an example for 2 bins with an un-trained neural network:
image
Before optimization, galaxies are more or less randomly assigned.

But after training on `riz` you can get results like this:
nn_riz_2
It successfully learns that you better build independent redshift bins. But see that there are some still some catastrophic outliers. 

What I didn't expect and find super interesting, is that sometimes the neural network finds original ways to bin the data:
riz_2
Funny enough, this second solution has higher S/N than the default tree-based solution that looks like the above one. Pretty cool.....

This is just a draft PR because I'm existed about these prelimary results :-) I've only done `riz` so far, and find a consistent +2 in S/N compared to the baseline, but without any tuning at all.

I also have all the tools to implement FoM optimization, but I ran out of time >.<

Thoughts/comments  welcome! I'm curious to know if we can expect to win a lot compared to tree classifcation....

### CODE REVIEW

**Comment by joezuntz** [comment]
This sounds really, cool, and will presumably have much wider utility - in particular it could make Hamiltonian samplers work for us, and they're one of the few methods that sampling people agree work on generic high-dimensional spaces.

**Comment by EiffL** [comment]
yes yes yes :-D actually, the uses are endless..... 
I didn't mention it in the above post.... but the following works:
```
H = jax.hessian(3x2pt_likelihood)(cosmo)
```
which means you can get second order derivative information! with that, I can fit a posterior in like just a few iterations of variational inference. Bye Bye MCMC :-D

But, related to this project, you could find by optimization things like the most robust photoz uncertainty parametrization, or like isolate the galaxies that are going to be messing up the most the analysis. 

In any case @joezuntz feel free to take a look at that code, I'm still in the API design phase for this package, any inputs/feedback on how to structure it would be super helpful.

**Comment by EiffL** [comment]
@joezuntz I'm trying to match my Fisher matrices to the ones from cosmosis, and I'm a bit puzzled, maybe you'll have some hints.

I'm looking at a simple 1 bin 3x2pt to start with. My data vectors and cov, agree perfectly with CCL:
image

image

But when I derive the constant covariance Gaussian Fisher matrix using autodiff and try to compare the fisher matrix to the cosmosis output, 2 things seem to happen:
  1) I don't seem to get the right amplitude
  2) the order of parameters seem to be reversed
Here is a plot where the left matrix is the cosmosis fisher, the right is mine(after reverting the order of the axes):
image

For 1) I think I have a possible explanation... I haven't implemented halofit yet so only using linear power spectrum, so I disabled it as well in the CCL code in the metrics file for comparison. But I'm not actually sure how firecrown+cosmosis works, where does firecrown get like the power spectrum prescription to use? I'm thinking the amplitude is wrong because maybe the derivatives are being computed with a different model.

But for 2), I tried  a bunch of things and it does seem I'm getting the Fisher matrix with parameters in the reverse order.... any idea why this might happen?

**Comment by EiffL** [comment]
actually I think I answered 1) in #6

**Comment by joezuntz** [comment]
For 2 - possibly the order of parameters in the `cosmosis` section of the yaml config needs to be the same as the order in the parameters section?

**Comment by EiffL** [comment]
Actually, they may be in the right order, but the matrices don't agree and my brain is imagining things....
I've tried to simplify the Fisher to one or 2 parameters, they appear to look more or less in the right order, and the Fisher out of cosmosis is stable with respect to the delta parameter, but different from the one I get analytically.
So either:
1) I'm not getting the right gradients by autodiff somehow.
2) I'm not doing the same computation as firecrown.

I've been looking into 2) and I guess it's related to #5 , I don't see where in firecrown the noise power spectrum would be added to the signal vector. Or should it not be?

**Comment by EiffL** [comment]
Ok, thanks to #5 I'm a little bit more clear on what's going, I'll go back and try to make sure I'm getting the same one parameter derivatives.

**Comment by EiffL** [comment]
:-| the plot thickens.....
I get perfect agreements on derivatives I perform manually on Omega_c with CCL witth the same delta as the one I use in cosmosis:

image
If I compute this 1 parameter Fisher manually, I get:
 - finite differencing CCL: 3081752.2
 - finite differencing JAX: 3081752.2
 - autodiff JAX: 3081600.5

But what should be the same computation with firecrown gives me: 2316659.9

Is firecrown-cosmosis really doing: 
image
with the covariance of the SACC file?

and if so, am I going crazy and messing my einsum expression:
```python
F = np.einsum('pa,pq,qb->ab', dmu, np.linalg.inv(C), dmu)
```
with dmu of shape [n_obs, n_params] ?

**Comment by joezuntz** [comment]
That's exactly what cosmosis is doing.  Although in cosmosis the step size is relative to the width of the parameter prior, so that might be different from what you're doing?

**Comment by joezuntz** [comment]
also cosmosis is doing a 5-point stencil - are you doing 3 points?

**Comment by EiffL** [comment]
Sorry, I didn't see  your  response right away as I was reading the cosmosis code ^^' 
Yeah, I'm only doing +/- delta for the derivative, but given that they alsso agree with the autodiff one,  I was quite confident that worked. 

I see a prior term in cosmosis, but it shouldn't be added to the Fisher if I understand  the config correctly, unless maybe firecrown adds a prior somewhere....

**Comment by joezuntz** [comment]
Yes, only Gaussian priors are added.  Uniform ones are ignored.

**Comment by EiffL** [comment]
I'm puzzled because for some parameters it seems to work fine, this is a 2 param only fisher:
image
where the red is mine, and the black is cosmosis. 

But Omega_c gives me problems

**Comment by EiffL** [comment]
ooooooohhh you know what I think I got it, the metrics script only saves one value for the bias, right? So I'm guessing Firecrown ends up using a fixed bias, whereas I have 1/g(a)

**Comment by EiffL** [comment]
Yep :-D That was it!!!!
Obviously the derivatives for the clustering tracers are different now that I have fixed the bias to the value cosmosis is getting in the config (green and yellow), compared to the derivatives obtained with CCL and the metrics code (blue) that has varying bias :
image

And now, I'm getting roughly the same values for the one parameter Omega_c fisher:
- JAX autograd :  2289511.2
- cosmosis : 2316659.9

^^' wow, thanks for the debug and not quiet options in cosmosis! if I hadn't seen that fixed bias value I might never have found the problem 

Many thanks for the help Joe!

**Comment by EiffL** [comment]
Tadaa! 
image
This is back to the settings of the challenge, so with 5 params, black is cosmosis, red is mine obtained by autodiff \o/

^^' what an adventure..... Fisher matrices are evil! this is bringing back traumatic memories from my first Fisher matrix back when I was doing a project with @anaisrassat during my PhD :-D 

Allright, but  as a bonus, I have testsed a few step sizes for comosis, and looks like the range `0.002-0.005` is given good results, but the  default `0.02` is way off, in the one bin 3x2pt case:
image
blue and green and 0.002, 0.005 respectively, red and black are 0.01,0.02 respectively.

**Comment by EiffL** [comment]
Anais :-D yes, yes, you did, and it came back to haunt me :-D Hope you are doing ok in you corner of the world! I wasn't sure you would see this ping, I'm very happy you did :-D

**Comment by anaisrassat** [comment]
It was a pleasure to see the ping, so thanks @EiffL for the mention! Good to see you are exchanging with @joezuntz, one of the first people I met in cosmology too :) 
My corner of the world is doing ok, despite the fact that we seem to have moved into a parallel universe a few months ago.
